keyboards/zsa/voyager/voyager.c

```diff
@@ -0,0 +1,359 @@
+
+bool process_record_kb(uint16_t keycode, keyrecord_t *record) {
+    if (!process_record_user(keycode, record)) {
+        return false;
+    }
+    switch (keycode) {
+#if !defined(VOYAGER_USER_LEDS)
+        case LED_LEVEL:
+            if (record->event.pressed) {
+                keyboard_config.led_level ^= 1;
+                eeconfig_update_kb(keyboard_config.raw);
+                if (keyboard_config.led_level) {
+                    layer_state_set_kb(layer_state);
+                } else {
+                    STATUS_LED_1(false);
+                    STATUS_LED_2(false);
+                    STATUS_LED_3(false);
+                    STATUS_LED_4(false);
+                }
+            }
+            break;
+#endif
+#ifdef RGB_MATRIX_ENABLE
+        case TOGGLE_LAYER_COLOR:
+            if (record->event.pressed) {
+                keyboard_config.disable_layer_led ^= 1;
+                if (keyboard_config.disable_layer_led) rgb_matrix_set_color_all(0, 0, 0);
+            }
+            break;
+        case RGB_TOG:
+            if (record->event.pressed) {
+                switch (rgb_matrix_get_flags()) {
+                    case LED_FLAG_ALL: {
+                        rgb_matrix_set_flags(LED_FLAG_NONE);
+                        rgb_matrix_set_color_all(0, 0, 0);
+                    } break;
+                    default: {
+                        rgb_matrix_set_flags(LED_FLAG_ALL);
+                    } break;
+                }
+            }
+            return false;
+#endif
+    }
+    return true;
+}
+
+void matrix_init_kb(void) {
+    keyboard_config.raw = eeconfig_read_kb();
+
+    if (!keyboard_config.led_level && !keyboard_config.led_level_res) {
+        keyboard_config.led_level     = true;
+        keyboard_config.led_level_res = 0b11;
+        eeconfig_update_kb(keyboard_config.raw);
+    }
+    matrix_init_user();
+}
+
+void eeconfig_init_kb(void) { // EEPROM is getting reset!
+    keyboard_config.raw           = 0;
+    keyboard_config.led_level     = true;
+    keyboard_config.led_level_res = 0b11;
+    eeconfig_update_kb(keyboard_config.raw);
+    eeconfig_init_user();
+}
+
+
+// The keyboard uses a dfu-util compatible custom bootloader.
+// Rather than implementing in core, just override relevant functions.
+
+#define APP_ADDRESS 0x08002000
+
+void bootloader_jump(void) {
+    // The ignition bootloader is checking for a high signal on A8 for 100ms when powering on the board.
+    // Setting both A8 and A9 high will charge the capacitor quickly.
+    // Setting A9 low before reset will cause the capacitor to discharge
+    // thus making the bootloder unlikely to trigger twice between power cycles.
+    setPinOutputPushPull(A9);
+    setPinOutputPushPull(A8);
+    writePinHigh(A9);
+    writePinHigh(A8);
+    wait_ms(500);
+    writePinLow(A9);
+
+    NVIC_SystemReset();
+}
+
+
+void mcu_reset(void) {
+    // When resetting the MCU, we want to jump to the application.
+    SCB->AIRCR = APP_ADDRESS & 0xFFFF;
```

Comment:
According to the Cortex-M4 programming manual, this AIRCR write actually does nothing:
![image](https://github.com/qmk/qmk_firmware/assets/616082/83119f49-cbc4-4e6d-ba3f-34c896e83f9e)